File: src/ai/provider-options.ts
import type { AiSettings } from "../settings.ts";
import { getModelCapability } from "./capability.ts";
import type { ProviderModelDefaults } from "../providers/config.ts";

type JSONValue = string | number | boolean | null | JSONValue[] | { [key: string]: JSONValue };

/**
 * Gemini 3 系（gemini-3 / gemini-3.1 / gemini-3.5 ...）かどうかを判定する。
 */
export function isGemini3Model(model: string): boolean {
  return /^gemini-3(\.|-|$)/.test(model);
}

/**
 * モデルの capability を考慮してプロバイダーオプションを構築する。
 * 選択されたモデルの reasoningCapability メタデータを参照し、
 * サポートされているオプションのみをプロトコルに送出する。
 * 非対応のフィールド（例: adaptive-only モデルへの budget）は送らない。
 *
 * @param settings 現在の AI 設定
 * @param _toolsEnabled 旧仕様の名残(DeepSeek thinking がツールと両立するため未使用。互換のため残置)
 * @param defaults 選択されたモデルの ProviderModelDefaults（reasoningCapability を含む）
 */
export function buildProviderOptions(
  settings: AiSettings,
  _toolsEnabled = false,
  defaults?: ProviderModelDefaults,
): Record<string, Record<string, JSONValue>> | undefined {
  const provider = settings.provider;
  const modelId = settings.model;

  // Use ephemeral reasoningCapability from settings if available, otherwise resolve from defaults
  const cap = settings.reasoningCapability ?? getModelCapability(provider, modelId, defaults);

  switch (provider) {
    case "openai": {
      if (!settings.openaiReasoningEffort) return undefined;
      const baseUrl = typeof settings.baseUrl === "string" ? settings.baseUrl.trim() : "";
      if (baseUrl.includes("api.platform.preferredai.jp")) return undefined;
      return {
        openai: {
          // GPT-5.1 以降 "minimal" は API から削除済み("none" が後継)。旧設定を救済する。
          reasoningEffort: settings.openaiReasoningEffort === "minimal" ? "none" : settings.openaiReasoningEffort,
          // "detailed" はモデルスナップショットによって拒否された実績があるため公式推奨の "auto" を使う
          reasoningSummary: "auto",
        },
      };
    }
    case "anthropic": {
      if (!cap) {
        // capability 不明のモデルには従来どおり effort 設定時のみ adaptive を送る
        if (!settings.anthropicThinkingEffort) return undefined;
        return {
          anthropic: {
            thinking: { type: "adaptive" },
            effort: settings.anthropicThinkingEffort,
          },
        };
      }
      if (cap.kind === "anthropic-adaptive") {
        // Opus 4.8 のような「adaptive だが無効化可」のモデルでは、無効時に
        // thinking を一切送らない(省略 = OFF。{type:"disabled"} は Fable 5 で 400)。
        if (cap.canDisable === true && settings.anthropicThinkingEnabled === false) {
          return undefined;
        }
        // Fable 5 のような常時 ON adaptive モデルは、effort を指定しない場合でも
        // thinking 自体はサーバ側で有効になるため providerOptions を省略する。
        if (cap.canDisable !== true && !settings.anthropicThinkingEffort) {
          return undefined;
        }
        // Never emit invented display:"detailed"; only emit display when metadata says "summarized"
        const thinking: Record<string, JSONValue> = { type: "adaptive" };
        if (cap.display === "summarized") {
          thinking.display = "summarized";
        }
        return {
          anthropic: {
            thinking,
            ...(settings.anthropicThinkingEffort && { effort: settings.anthropicThinkingEffort }),
          },
        };
      }
      // Budget thinking supports explicit disable. Only enabled mode needs a budget.
      if (settings.anthropicThinkingEnabled === false) {
        return { anthropic: { thinking: { type: "disabled" } } };
      }
      if (settings.anthropicThinkingBudget == null) return undefined;
      return {
        anthropic: {
          thinking: {
            type: "enabled",
            budgetTokens: settings.anthropicThinkingBudget,
          },
        },
      };
    }
    case "deepseek": {
      // DeepSeek V3.2 以降、thinking モードはツール呼び出しと両立する。
      // ON/OFF は設定のみで決める(ツール有効時の強制 OFF は V3 時代の制約)。
      if (settings.deepseekThinkingEnabled === false) {
        return { deepseek: { thinking: { type: "disabled" } } };
      }
      const options: Record<string, JSONValue> = {
        thinking: { type: "enabled" },
      };
      // API の有効値は "high" / "max" のみ。それ以外はサーバ既定(high)に任せる。
      if (settings.deepseekReasoningEffort === "high" || settings.deepseekReasoningEffort === "max") {
        options.reasoningEffort = settings.deepseekReasoningEffort;
      }
      return { deepseek: options };
    }
    case "codex": {
      const effort = settings.openaiReasoningEffort === "minimal" ? "none" : settings.openaiReasoningEffort;
      return {
        openai: {
          ...(effort && { reasoningEffort: effort }),
          reasoningSummary: "auto",
          // Codex backend はサーバ側に会話状態を持たない。store:false と
          // encrypted reasoning の往復が無いとマルチターンのツール呼び出しが
          // 2 ターン目に 400 ("reasoning without its required following item") になる。
          store: false,
          include: ["reasoning.encrypted_content"],
        },
      };
    }
    case "github-copilot": {
      // Use the same effective capability that drove the UI/model routing. The
      // runtime places the authenticated Copilot cache override on settings.
      if (cap?.kind === "anthropic-adaptive") {
        if (!settings.anthropicThinkingEffort) return undefined;
        const thinking: Record<string, JSONValue> = { type: "adaptive" };
        if (cap.display === "summarized") thinking.display = "summarized";
        return { anthropic: { thinking, effort: settings.anthropicThinkingEffort } };
      }
      if (cap?.kind === "anthropic-budget") {
        if (settings.anthropicThinkingEnabled === false) {
          return { anthropic: { thinking: { type: "disabled" } } };
        }
        if (settings.anthropicThinkingBudget == null) return undefined;
        return { anthropic: { thinking: { type: "enabled", budgetTokens: settings.anthropicThinkingBudget } } };
      }
      if (cap?.kind === "openai") {
        if (!settings.openaiReasoningEffort) return undefined;
        return {
          openai: {
            reasoningEffort: settings.openaiReasoningEffort,
            reasoningSummary: "auto",
            include: ["reasoning.encrypted_content"],
          },
        };
      }
      return undefined;
    }
    case "sakura":
      // Sakura の Responses API はマルチターン非対応のため Chat Completions を使う。
      // store / strictJsonSchema / reasoningSummary は Responses 専用なので送らない。
      return {
        openai: {
          parallelToolCalls: false,
        },
      };
    case "google": {
      if (!isGemini3Model(settings.model)) return undefined;
      // thinkingLevel の有効値はモデル別(gemini-3.1-pro 系は "minimal" 非対応)。
      // capability の supportedEfforts に無い値は "minimal"→"low" に格上げし、
      // それでも不一致なら送らずサーバ既定に任せる。
      let level = settings.googleThinkingLevel;
      const supported = cap?.kind === "google" ? cap.supportedEfforts : undefined;
      if (level && Array.isArray(supported) && !supported.includes(level)) {
        const fallback = level === "minimal" ? "low" : undefined;
        level = fallback && supported.includes(fallback) ? fallback : undefined;
      }
      return {
        google: {
          thinkingConfig: {
            includeThoughts: true,
            ...(level ? { thinkingLevel: level } : {}),
          },
        },
      };
    }
    default:
      return undefined;
  }
}

/**
 * debugFetch が独自のリトライ（カスタムバックオフ）を行うプロバイダでは
 * AI SDK の標準リトライを無効化し、二重リトライによる過剰な待ち時間を防ぐ。
 */
export function buildRetryOption(settings: AiSettings): { maxRetries?: number } {
  if (
    settings.provider === "opencode" ||
    settings.provider === "sakura" ||
    settings.provider === "codex" ||
    settings.provider === "github-copilot"
  ) {
    return { maxRetries: 0 };
  }
  return {};
}

/**
 * AI 呼び出しのエラーメッセージをユーザー向けに整形する。
 */
export function formatAiErrorMessage(error: unknown): string {
  const raw = error instanceof Error ? error.message : String(error);
  if (
    /upstream request failed|upstream error|upstream unavailable|overloaded|temporarily unavailable|service unavailable|rate[- ]?limit|too many requests|throttl/i.test(
      raw,
    )
  ) {
    return `${raw}\n\n時間をおいて再度送信してください。`;
  }
  return raw;
}

File: frontend-rs/src/ai/provider_options.rs
//! モデルの capability を考慮してプロバイダーオプションを構築する。
//! 選択されたモデルの reasoningCapability メタデータを参照し、
//! サポートされているオプションのみをプロトコルに送出する。
//!
//! TypeScript `provider-options.ts` の Rust 移植。

use serde_json::{Map, Value};

use super::capability::{self, ReasoningCapability};

/// モデルの capability を考慮してプロバイダーオプションを構築する。
///
/// `settings` は JSON object として現在の AI 設定全体を含む。
/// `defaults_capability` は選択されたモデルの ProviderModelDefaults 由来の
/// reasoningCapability（存在すれば）。
///
/// 戻り値はプロバイダーオプションの JSON object、または不要な場合は None。
pub fn build_provider_options(
    settings: &Value,
    defaults_capability: Option<&ReasoningCapability>,
) -> Option<Value> {
    let provider = settings
        .get("provider")
        .and_then(Value::as_str)
        .unwrap_or("");
    let model_id = settings
        .get("model")
        .and_then(Value::as_str)
        .unwrap_or("");

    let cap = settings
        .get("reasoningCapability")
        .and_then(|v| serde_json::from_value::<ReasoningCapability>(v.clone()).ok())
        .or_else(|| capability::get_model_capability(provider, model_id, defaults_capability));

    match provider {
        "openai" => {
            let effort = settings
                .get("openaiReasoningEffort")
                .and_then(Value::as_str);
            let Some(effort) = effort else { return None };
            let base_url = settings
                .get("baseUrl")
                .and_then(Value::as_str)
                .unwrap_or("");
            if base_url.contains("api.platform.preferredai.jp") {
                return None;
            }
            // GPT-5.1 以降 "minimal" は API から削除済み（"none" が後継）。旧設定を救済する。
            let effort = if effort == "minimal" { "none" } else { effort };
            let mut opts = Map::new();
            opts.insert("reasoningEffort".into(), Value::String(effort.into()));
            opts.insert("reasoningSummary".into(), Value::String("auto".into()));
            let mut result = Map::new();
            result.insert("openai".into(), Value::Object(opts));
            Some(Value::Object(result))
        }
        "anthropic" => {
            let cap = cap.as_ref();
            if cap.is_none() {
                let effort = settings
                    .get("anthropicThinkingEffort")
                    .and_then(Value::as_str);
                let Some(effort) = effort else { return None };
                let mut thinking = Map::new();
                thinking.insert("type".into(), Value::String("adaptive".into()));
                let mut opts = Map::new();
                opts.insert("thinking".into(), Value::Object(thinking));
                opts.insert("effort".into(), Value::String(effort.into()));
                let mut result = Map::new();
                result.insert("anthropic".into(), Value::Object(opts));
                return Some(Value::Object(result));
            }
            let cap = cap.unwrap();
            if cap.kind == "anthropic-adaptive" {
                if cap.can_disable
                    && settings
                        .get("anthropicThinkingEnabled")
                        .and_then(Value::as_bool)
                        == Some(false)
                {
                    return None;
                }
                if !cap.can_disable
                    && settings.get("anthropicThinkingEffort").is_none()
                {
                    return None;
                }
                let mut thinking = Map::new();
                thinking.insert("type".into(), Value::String("adaptive".into()));
                if cap.display.as_deref() == Some("summarized") {
                    thinking.insert(
                        "display".into(),
                        Value::String("summarized".into()),
                    );
                }
                let mut opts = Map::new();
                opts.insert("thinking".into(), Value::Object(thinking));
                if let Some(effort) =
                    settings.get("anthropicThinkingEffort").and_then(Value::as_str)
                {
                    opts.insert("effort".into(), Value::String(effort.into()));
                }
                let mut result = Map::new();
                result.insert("anthropic".into(), Value::Object(opts));
                Some(Value::Object(result))
            } else {
                // Budget thinking
                if settings
                    .get("anthropicThinkingEnabled")
                    .and_then(Value::as_bool)
                    == Some(false)
                {
                    let mut thinking = Map::new();
                    thinking.insert("type".into(), Value::String("disabled".into()));
                    let mut opts = Map::new();
                    opts.insert("thinking".into(), Value::Object(thinking));
                    let mut result = Map::new();
                    result.insert("anthropic".into(), Value::Object(opts));
                    return Some(Value::Object(result));
                }
                let budget = settings
                    .get("anthropicThinkingBudget")
                    .and_then(Value::as_u64);
                let Some(budget) = budget else { return None };
                let mut thinking = Map::new();
                thinking.insert("type".into(), Value::String("enabled".into()));
                thinking.insert(
                    "budgetTokens".into(),
                    Value::Number(budget.into()),
                );
                let mut opts = Map::new();
                opts.insert("thinking".into(), Value::Object(thinking));
                let mut result = Map::new();
                result.insert("anthropic".into(), Value::Object(opts));
                Some(Value::Object(result))
            }
        }
        "deepseek" => {
            let thinking_disabled = settings
                .get("deepseekThinkingEnabled")
                .and_then(Value::as_bool)
                == Some(false);
            if thinking_disabled && !capability::is_deepseek_v4_model(model_id) {
                let mut thinking = Map::new();
                thinking.insert("type".into(), Value::String("disabled".into()));
                let mut opts = Map::new();
                opts.insert("thinking".into(), Value::Object(thinking));
                let mut result = Map::new();
                result.insert("deepseek".into(), Value::Object(opts));
                return Some(Value::Object(result));
            }
            let mut thinking = Map::new();
            thinking.insert("type".into(), Value::String("enabled".into()));
            let mut opts = Map::new();
            opts.insert("thinking".into(), Value::Object(thinking));
            if let Some(effort) = settings
                .get("deepseekReasoningEffort")
                .and_then(Value::as_str)
            {
                if effort == "high" || effort == "max" {
                    opts.insert(
                        "reasoningEffort".into(),
                        Value::String(effort.into()),
                    );
                }
            }
            let mut result = Map::new();
            result.insert("deepseek".into(), Value::Object(opts));
            Some(Value::Object(result))
        }
        "codex" => {
            let effort = settings
                .get("openaiReasoningEffort")
                .and_then(Value::as_str);
            let effort = effort.map(|e| if e == "minimal" { "none" } else { e });
            let mut opts = Map::new();
            if let Some(effort) = effort {
                opts.insert(
                    "reasoningEffort".into(),
                    Value::String(effort.into()),
                );
            }
            opts.insert("reasoningSummary".into(), Value::String("auto".into()));
            opts.insert("store".into(), Value::Bool(false));
            opts.insert(
                "include".into(),
                Value::Array(vec![Value::String(
                    "reasoning.encrypted_content".into(),
                )]),
            );
            let mut result = Map::new();
            result.insert("openai".into(), Value::Object(opts));
            Some(Value::Object(result))
        }
        "github-copilot" => {
            let cap = cap.as_ref();
            match cap.map(|c| c.kind.as_str()) {
                Some("anthropic-adaptive") => {
                    let effort = settings
                        .get("anthropicThinkingEffort")
                        .and_then(Value::as_str);
                    let Some(effort) = effort else { return None };
                    let mut thinking = Map::new();
                    thinking.insert(
                        "type".into(),
                        Value::String("adaptive".into()),
                    );
                    if cap.unwrap().display.as_deref() == Some("summarized") {
                        thinking.insert(
                            "display".into(),
                            Value::String("summarized".into()),
                        );
                    }
                    let mut opts = Map::new();
                    opts.insert("thinking".into(), Value::Object(thinking));
                    opts.insert("effort".into(), Value::String(effort.into()));
                    let mut result = Map::new();
                    result.insert("anthropic".into(), Value::Object(opts));
                    Some(Value::Object(result))
                }
                Some("anthropic-budget") => {
                    if settings
                        .get("anthropicThinkingEnabled")
                        .and_then(Value::as_bool)
                        == Some(false)
                    {
                        let mut thinking = Map::new();
                        thinking.insert(
                            "type".into(),
                            Value::String("disabled".into()),
                        );
                        let mut opts = Map::new();
                        opts.insert("thinking".into(), Value::Object(thinking));
                        let mut result = Map::new();
                        result.insert("anthropic".into(), Value::Object(opts));
                        return Some(Value::Object(result));
                    }
                    let budget = settings
                        .get("anthropicThinkingBudget")
                        .and_then(Value::as_u64);
                    let Some(budget) = budget else { return None };
                    let mut thinking = Map::new();
                    thinking.insert(
                        "type".into(),
                        Value::String("enabled".into()),
                    );
                    thinking.insert(
                        "budgetTokens".into(),
                        Value::Number(budget.into()),
                    );
                    let mut opts = Map::new();
                    opts.insert("thinking".into(), Value::Object(thinking));
                    let mut result = Map::new();
                    result.insert("anthropic".into(), Value::Object(opts));
                    Some(Value::Object(result))
                }
                Some("openai") => {
                    let effort = settings
                        .get("openaiReasoningEffort")
                        .and_then(Value::as_str);
                    let Some(effort) = effort else { return None };
                    let mut opts = Map::new();
                    opts.insert(
                        "reasoningEffort".into(),
                        Value::String(effort.into()),
                    );
                    opts.insert(
                        "reasoningSummary".into(),
                        Value::String("auto".into()),
                    );
                    opts.insert(
                        "include".into(),
                        Value::Array(vec![Value::String(
                            "reasoning.encrypted_content".into(),
                        )]),
                    );
                    let mut result = Map::new();
                    result.insert("openai".into(), Value::Object(opts));
                    Some(Value::Object(result))
                }
                _ => None,
            }
        }
        "sakura" => {
            let mut opts = Map::new();
            opts.insert(
                "parallelToolCalls".into(),
                Value::Bool(false),
            );
            let mut result = Map::new();
            result.insert("openai".into(), Value::Object(opts));
            Some(Value::Object(result))
        }
        "google" => {
            if !capability::is_gemini3_model(model_id) {
                return None;
            }
            let level = settings
                .get("googleThinkingLevel")
                .and_then(Value::as_str);
            let supported = cap
                .as_ref()
                .filter(|c| c.kind == "google")
                .map(|c| &c.supported_efforts);

            let mut level = level.map(|l| l.to_string());
            if let (Some(l), Some(supported)) = (&level, supported) {
                if !supported.contains(l) {
                    let fallback = if l == "minimal" {
                        Some("low".to_string())
                    } else {
                        None
                    };
                    level = fallback.filter(|f| supported.contains(f));
                }
            }

            let mut thinking_config = Map::new();
            thinking_config.insert(
                "includeThoughts".into(),
                Value::Bool(true),
            );
            if let Some(l) = level {
                thinking_config
                    .insert("thinkingLevel".into(), Value::String(l));
            }
            let mut opts = Map::new();
            opts.insert(
                "thinkingConfig".into(),
                Value::Object(thinking_config),
            );
            let mut result = Map::new();
            result.insert("google".into(), Value::Object(opts));
            Some(Value::Object(result))
        }
        _ => None,
    }
}

/// debugFetch が独自のリトライを行うプロバイダでは AI SDK の標準リトライを無効化し、
/// 二重リトライによる過剰な待ち時間を防ぐ。
pub fn build_retry_option(provider: &str) -> Option<u32> {
    match provider {
        "opencode" | "sakura" | "codex" | "github-copilot" => Some(0),
        _ => None,
    }
}

/// AI 呼び出しのエラーメッセージをユーザー向けに整形する。
pub fn format_ai_error_message(error: &str) -> String {
    let lower = error.to_lowercase();
    if lower.contains("upstream request failed")
        || lower.contains("upstream error")
        || lower.contains("upstream unavailable")
        || lower.contains("overloaded")
        || lower.contains("temporarily unavailable")
        || lower.contains("service unavailable")
        || lower.contains("rate limit")
        || lower.contains("rate-limit")
        || lower.contains("too many requests")
        || lower.contains("throttl")
    {
        format!("{error}\n\n時間をおいて再度送信してください。")
    } else {
        error.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn openai_basic_options() {
        let settings = json!({
            "provider": "openai",
            "model": "gpt-5.1",
            "openaiReasoningEffort": "medium",
            "baseUrl": ""
        });
        let result = build_provider_options(&settings, None);
        assert!(result.is_some());
        let opts = result.unwrap();
        assert_eq!(
            opts["openai"]["reasoningEffort"].as_str().unwrap(),
            "medium"
        );
    }

    #[test]
    fn openai_minimal_upgraded_to_none() {
        let settings = json!({
            "provider": "openai",
            "model": "gpt-5.1",
            "openaiReasoningEffort": "minimal",
            "baseUrl": ""
        });
        let result = build_provider_options(&settings, None);
        assert_eq!(
            result.unwrap()["openai"]["reasoningEffort"]
                .as_str()
                .unwrap(),
            "none"
        );
    }

    #[test]
    fn deepseek_thinking_disabled() {
        let settings = json!({
            "provider": "deepseek",
            "model": "deepseek-v3",
            "deepseekThinkingEnabled": false
        });
        let result = build_provider_options(&settings, None);
        assert!(result.is_some());
        let opts = result.unwrap();
        assert_eq!(
            opts["deepseek"]["thinking"]["type"].as_str().unwrap(),
            "disabled"
        );
    }

    #[test]
    fn deepseek_v4_never_disabled() {
        let settings = json!({
            "provider": "deepseek",
            "model": "deepseek-v4-0324",
            "deepseekThinkingEnabled": false
        });
        let result = build_provider_options(&settings, None);
        let opts = result.unwrap();
        assert_eq!(
            opts["deepseek"]["thinking"]["type"].as_str().unwrap(),
            "enabled"
        );
    }

    #[test]
    fn sakura_disables_parallel_tool_calls() {
        let settings = json!({
            "provider": "sakura",
            "model": "some-model"
        });
        let result = build_provider_options(&settings, None);
        assert_eq!(
            result.unwrap()["openai"]["parallelToolCalls"]
                .as_bool()
                .unwrap(),
            false
        );
    }
}
